test_runner/fixtures/utils.py

```diff
@@ -604,3 +615,34 @@ def human_bytes(amt: float) -> str:
         amt = amt / 1024
 
     raise RuntimeError("unreachable")
+
+
+def all_pairs_component_versions():
+    """
+    This function generates all the pairs of old or new versions of the Neon components
+    E.g. Pairs(storage_controller='old', storage_broker='new', compute='new', safekeeper='old', pageserver='old')
+    then it returns a dictionary with argnames, argvalues and ids
+    ids is a sequence of letters where n means the new version of the component and o means the old version
+    """
+    argvalues, ids = [], []
+    all_new = False
+    for pair in AllPairs(
+        OrderedDict({component: ["new", "old"] for component in COMPONENT_BINARIES.keys()})
+    ):
+        have_old = have_new = False
+        argvalues.append(pair)
+        cur_id = []
+        for component in COMPONENT_BINARIES.keys():
+            if getattr(pair, component) == "new":
+                cur_id.append("n")
+                have_new = True
+            else:
+                cur_id.append("o")
+                have_old = True
+        ids.append("combination_" + "".join(cur_id))
+        all_new = all_new or (have_new and not have_old)
+        assert (
+            have_new
+        ), f"the wrong combination {pair} is generated, we don't expect old versions only"
+    assert all_new, f"the combination of all-new versions is not generated, please check all-pairs setup. {argvalues}"
```

Comment:
`AllPairs` accepts `filter_func` to filter out unwanted combinations (see the [README](https://github.com/thombashi/allpairspy?tab=readme-ov-file)).
I would filter out all-new and all-old combinations and then add all-new to the returning result.

Reply:
If I use the filter function it generates one more combination.
So without the filter, we had 6 combinations, but if we filter out the all-new combination we still get 6, so it will be 7 combinations then if we add all-new one by hand